lib/contacts.py

```diff
@@ -169,6 +169,7 @@ def save(self):
         d = self._save(self.data, v1_too = False)  # Note: set v1_too = True if you want to save to v1 so older EC wallets can also see the updated contacts
         for k,v in d.items():
             self.storage.put(k, v)  # "contacts2", "contacts" are the two expected keys
+        self.storage.write()
```

Comment:
Writing to storage is expensive potentially because it causes the code to re-encrypt the wallet (if the wallet is encrypted). 

Also: `contacts.save` could be called anywhere, such as in the context of other code that is saving a bunch of things to the wallet.


It should not be a decision made in this function to call storage.write but rather by the caller who has more information about whether this is a mass save or a one-off.  

When I rewrote this class this month, I was aware of that fact. This code intentionally does not call storage.write here.  In other classes such as `paymentrequest.py` we do not call storage.write either.

Look in wallet.py in the various places `contacts.save` is called.  It is called in the context of saving a bunch of different things, one after another.  If each of those called functions called `storage.write`, it would negatively impact performance.  

Please remove this call.

(If we want to make sure contacts are actually saved to file after modification -- we are free to call storage.write ourselves in client code, gui code, etc -- this is not the place for it in light of the performance considerations.)

Reply:
Hey if you want can:

- add an arg to the function `write=False` (defaults False)
- if True, then can call `storage.write()`

Sound ok?

Reply:
Oh wait -- what the hell did I just grep?  Contacts.save is only called within this very class! Lol!

Yeah.. this is fine.. I think.

I am confused.  I think I accidentally grepped `cashacct.save`..and confused myself.

I still am nervouse about calling `storage.write` here but I guess it's ok. Hmm.

Reply:
Yeah really I'd rather the unconditional `storage.write` call be removed.  On Qt UI it's pretty hard to get the wallet to *not* save to disk.. even if the app is quickly quit it always saves the wallet anyway.  Same on iOS.  I dunno -- seems like it takes care of itself eventually anyway...

Reply:
On Qt desktop we have huge wallets sometimes (like 80MB+) and every darned save call is another UI lag that doesn't need to happen.. hence my reticence here.

Reply:
OK, I've moved the write call to the GUI code. On Android this definitely does not take care of itself: when you install a new build of the app from the IDE, it kills the existing process without giving it any warning.

Reply:
You could feel free to pass a bool write=True as well. That pattern exists in various bits of the code here.  But your call.

Sorry to be difficult about this -- I've tried to tune performance of the desktop app. There was a time not long ago when it would issue multiple `sotrage.write` calls one after another after another hanging the UI on larger wallets (like CashShuffle).  So I have PTSD from that.  Thanks a lot for changing it. 👍

Reply:
No problem, I understand. 

I decided not to add the extra argument, because then I'd have to add it to `add` and `remove` as well, which are the functions the GUI actually calls.